Sources/Basic/Result.swift

```diff
@@ -0,0 +1,86 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// An simple enum which is either a value or an error.
+/// It can be used for error handling in situations where try catch is
+/// problematic to use, for eg: asynchronous APIs.
+public enum Result<Value, ErrorType: Swift.Error> {
+    /// Indicates success with value in the associated object.
+    case success(Value)
+
+    /// Indicates failure with error inside the associated object.
+    case failure(ErrorType)
+
+    /// Initialiser for value.
+    public init(_ value: Value) {
+        self = .success(value)
+    }
+
+    /// Initialiser for error.
+    public init(_ error: ErrorType) {
+        self = .failure(error)
+    }
+
+    /// Initialise with something that can throw.
+    public init(_ body: () throws -> Value) {
+        do {
+            self = .success(try body())
+        } catch {
+            self = .failure(error as! ErrorType)
+        }
+    }
+
+    /// Initialise with something that can throw.
+    public init(wrap body: @autoclosure () throws -> Value) {
+        self.init(body)
+    }
+
+    /// Get the value if success else throw the saved error.
+    public func dematerialize() throws -> Value {
+        switch self {
+        case .success(let value):
+            return value
+        case .failure(let error):
+            throw error
+        }
+    }
+}
+
+extension Result: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .success(let value):
+            return "\(value)"
```

Comment:
I'd probably prefer to show `Result()` here... I find it confusing when somethings description makes it looks like a different type.